Pipfile

```diff
@@ -15,9 +15,11 @@ python-decouple = "*"
 sh = "*"
 django = ">=2.2,<3"
 typing-extensions = "*"
+pympi-ling = "*"
+requests = "*"
+divvunspell-python = {ref = "3aa656d36a2fa1ede90e53c320b2095c05cf7c05", git = "https://github.com/divvun/divvunspell-sdk-python"}
 
 [dev-packages]
-black = "*"
```

Comment:
> Based on this thread,
pip install -U virtualenv of all things seemed to have let it install for me, but leaves me even more confused than before.

wat

Reply:
wat indeed. I am no longer able to reproduce the issue.

The error that `pipenv` produced was:

    raise VersionConflict(dist, req)
    pkg_resources.VersionConflict: (importlib-metadata 3.4.0 (~/.local/share/virtualenvs/recording-validation-interface-j9JF82f7/lib/python3.7/site-packages), Requirement.parse('importlib-metadata<3,>=0.12; python_version < "3.8"'))

which was confusing because while one `importlib-metadata` constraint was clearly coming from a `pytest` dependency, the `importlib-metadata<3` constraint seemingly came out of nowhere. It turns out it was coming from the implicit `virtualenv` dependency, which has such a constraint:

    ~/Library/Python/3.9/lib/python/site-packages/virtualenv-20.0.33.dist-info$ grep importlib $PWD/METADATA
    Requires-Dist: importlib-metadata (<3,>=0.12) ; python_version < "3.8"

So some combination of old/incompatible versions of `pipenv`/`virtualenv` and installing a `Pipfile` for 3.7, using a `pipenv` that may have been running on 3.9, triggered a bug where `virtualenv` had constraints that `pipenv` did not consider and ended up being unsatisfiable.

Reply:
To follow up: I have now reproduced the issue. My earlier attempts to do so were failing because, in trying to make a clean repro, I was `pipenv -rm`ing and then forgetting to add the `--dev` flag on install. No `pytest` ⇒ no `VersionConflict`.

With pipenv 2020.11.15 running under Python 3.9.1, and virtualenv version 20.0.33, 20.0.34, 20.0.35, 20.1.0, or 20.2.0 installed, `pipenv install -v black==20.8b1` fails inside a `pipenv shell` with Python 3.7 as the active Python.

Upgrading the *user* install of `virtualenv`, for *Python 3.9*, to 20.2.1 fixed the issue. Presumably later versions of `virtualenv` would too. There must have been some bug where the dependency constraints from the version of `virtualenv` that `pipenv` is running under were leaking into the lock calculation. Python code running inside the pipenv environment (this is *not* where `pipenv` itself runs, because it constructs that environment) only sees `/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/venv`.

Reply:
ay dios mio

![](https://media.giphy.com/media/JogZYXkQRpuGMKba7l/giphy.gif)

Thank you for the sluething!